### DISCUSSION TITLE
How is the "Average Force" value calculated???

### DISCUSSION THREAD

**Discussion by molegod** [empriselab/RCareWorld#58] [Help - Programming] [answered]
For some reason, if we have the sponge, and then place it directly on the human, which causes the body to be painted, there is still 0 force registering at all. How exactly is this value calculated? As it does not seem to be consistent - pushing the sponge in from the side also does not register any force.

**Comment by YoruCathy** [comment]
Does the sponge go inside the human body? If that’s the case there won’t be any forces showing up. Also only the bottom of the sponge counts as valid contact.

**Comment by utkudotdev** [comment]
That makes sense. We still sometimes see the force values we get on the environment's `instance_channel` being very different from those shown on the screen.

Also, we never get a force score because our code happens before the scoring (like in the sample files). How should we get the force score?

**Comment by YoruCathy** [comment, marked as answer]
Sometimes the force values might be laggy. That could be the issue. You can save the force values to some files, and then run the force score on the saved files.

**Reply by martinschonger** [reply]
I'm facing the same issue as OP - the `GetForce` method *never* gives anything but 0 (even averaged; also tried `GetEffectiveForceProportion`), no matter how slowly I lower the sponge (gripped by the robot) down onto the mannequin. Could you please provide a MWE that results in nonzero force reading there? I'm probably just missing a small detail..
The 'Average Force' value in the Unity player does usually take some nonzero value (inconsistently so, but at least nonzero).

**Reply by YoruCathy** [reply]
You can try printing out `sponge.data` and see if the forces in that dict matches with the one in the Unity player window

**Reply by martinschonger** [reply]
Thanks for your reply. When `sponge.data['forces']` is available, it matches the 'Average Force' value in the player. So, I'm assuming this value is the average over some time delta. How is this delta determined? Is it dependent on `env.data["fixed_delta_time"]`?

Another related question: On which parameters does the number of values available in `sponge.data['forces']` depend exactly? Depending on `env.data["fixed_delta_time"]`, sometimes after `env.step(35)` there is one value available, sometimes two; sometimes only after `env.step(80)` there is one value available (and after e.g. `env.step(70)` the key `forces` may not be set at all).

**Reply by martinschonger** [reply]
@YoruCathy I believe the question in the title of this discussion is still open. Is there any chance you could provide the Unity project file, so I can check the (average) force calculation directly? I would specifically like to debug `SpongeAttr.cs` and `ForceReader.cs`.
Another question related to object collisions: Which version of the ObiSolver are you using?

**Comment by thisjustin123** [comment]
Hello! Sorry to revive this thread, but I've noticed a bunch of bugs with the sponge force reading on the Unity side. I'll try to fix these up and push a fix soon. I've fixed them locally, but just need some time to make sure they're ready to ship.

Sorry about the delay in general on fixes. There are some logistical issues preventing bug fixes from coming out as quickly as we'd like.

**Reply by thisjustin123** [reply]
Working on as an extension to #131
